### PR TITLE
feat: support frame delays

### LINKED ISSUE #1338: GIF/WEBP animation duration
Hi, we are using a custom Sharp fork with a GIF and WEBP support.

However, we find out that some source images have various delays between images.
Unfortunately, we are not able to keep these delays intact as VIPS provides only one (probably the first) value.

In our example, the source GIF have 4 frames, where delay after each of first 3 is 500ms and then 2500ms after the last one. Then, the animation repeats.

When we process it with VIPS, the duration of animation is 2000ms instead of 4000ms as VIPS tell us that `gif-delay` is 50;

### PR DESCRIPTION
**Pull request by deftomat**
Support variable frame delays between frames.

Works for WEBP and GIF.

closes #1338

### CODE REVIEW

File: libvips/foreign/gifload.c [hunk 1]
```diff
@@ -575,19 +572,17 @@ vips_foreign_load_gif_set_header( VipsForeignLoadGif *gif, VipsImage *image )
...
 
-{
-	int array[10];
-	int i;
+  if ( gif->delays ) {
+	  vips_image_set_int( image, "gif-delay", VIPS_RINT( gif->delays[0] / 10.0 ) );
+	  vips_image_set_array_int( image, "delay", gif->delays, gif->n_pages );
+  } else {
+	  vips_image_set_int( image, "gif-delay", 4 );
+  }
```
Comment (deftomat) [inline libvips/foreign/gifload.c]: I'm not sure why it is aligned like this. Looks OK in editor 😞 

Maybe a TABS vs SPACES issue 🙂

**Comment by deftomat** [comment]
Hopefully reformatted correctly.

**Comment by jcupitt** [comment]
This looks great @deftomat! 

Sorry for the delay, I'll read this over today.

File: libvips/foreign/vips2webp.c [hunk 2]
```diff
@@ -348,7 +373,17 @@ write_webp_anim( VipsWebPWrite *write, VipsImage *image, int page_height )
 
 		WebPPictureFree( &pic );
 
-		timestamp_ms += delay;
+		int page_index = top / page_height;
```
Comment (jcupitt) [inline libvips/foreign/vips2webp.c]: In-block declarations are a C99 feature and libvips is (still) C89. Could you move the declaration to the start of the for().

File: libvips/foreign/webp2vips.c [hunk 3]
```diff
@@ -470,21 +470,32 @@ read_header( Read *read, VipsImage *out )
...
 
-		/* We must get the first frame to get the delay. Frames number
-		 * from 1.
-		 */
-		if( WebPDemuxGetFrame( read->demux, 1, &iter ) ) {
-			read->delay = iter.duration;
+		if ( read->frame_count > 1) {
+			read->delays = (int *) g_malloc( read->frame_count * sizeof(int) );
+
+			for( int i = 0; i < read->frame_count; i++ ) {
```
Comment (jcupitt) [inline libvips/foreign/webp2vips.c]: decls in for() are not in C89, could this move to the top of the block.

File: libvips/foreign/webp2vips.c [hunk 4]
```diff
@@ -470,21 +470,32 @@ read_header( Read *read, VipsImage *out )
 		vips_image_set_int( out, 
 			VIPS_META_PAGE_HEIGHT, read->frame_height );
 
-		/* We must get the first frame to get the delay. Frames number
-		 * from 1.
-		 */
-		if( WebPDemuxGetFrame( read->demux, 1, &iter ) ) {
-			read->delay = iter.duration;
+		if ( read->frame_count > 1) {
+			read->delays = (int *) g_malloc( read->frame_count * sizeof(int) );
```
Comment (jcupitt) [inline libvips/foreign/webp2vips.c]: You could use `g_malloc0()` here and make your loop below a little simpler.

File: libvips/foreign/webp2vips.c [hunk 5]
```diff
@@ -470,21 +470,32 @@ read_header( Read *read, VipsImage *out )
...
 
 #ifdef DEBUG
-			printf( "webp2vips: duration = %d\n", read->delay );
+		for( int i = 0; i < read->frame_count; i++ ) {
+			printf( "webp2vips: frame = %d; duration = %d\n", i + 1, read->delays[i] );
+		}
 #endif /*DEBUG*/
 
+			vips_image_set_array_int( out, "delay", read->delays, read->frame_count );
+			g_free( read->delays );
```
Comment (jcupitt) [inline libvips/foreign/webp2vips.c]: This will produce use-after-free errors. This should probably be a `VIPS_FREE()` in `read_free()`.

Comment (jcupitt) [reply libvips/foreign/webp2vips.c]: Use after free when you access the array on the next line, I mean.

**Comment by jcupitt** [comment]
I didn't mean to make you do loads more work -- I'm happy to merge and then make these changes myself if that's easier for you. Let me know.

**Comment by deftomat** [comment]
I've also added getter and setter for an array of integers: https://github.com/libvips/libvips/pull/1346/commits/1b5468465057c19282d71f345dfdf163436462fb

File: libvips/foreign/gifload.c [hunk 6]
```diff
@@ -521,11 +519,10 @@ vips_foreign_load_gif_scan_extension( VipsForeignLoadGif *gif )
 				gif->has_transparency = TRUE;
 			}
 
-			if( !gif->has_delay ) { 
-				VIPS_DEBUG_MSG( "gifload: has delay\n" ); 
-				gif->has_delay = TRUE;
-				gif->delay = extension[2] | (extension[3] << 8);
+			if( gif->n_pages % 64 == 0 ) {
+				gif->delays = (int *) g_realloc( gif->delays, (gif->n_pages + 64) * sizeof(int) );
```
[the remark this answers is not in the document]
Comment (deftomat) [reply libvips/foreign/gifload.c]: I'm not exactly sure what do you mean. We definitely need to depend on `n_pages` as thats the only indication of which frame is currently processed. 

I'm not sure if it is important if `n_pages` increase by one or not for a reallocation as `delay` array will be always bigger than `n_pages`. The only issue which I realize right now is that we rather need something like `g_realloc0()` which probably doesn't exists. 😞

**Comment by deftomat** [comment]
I fixed it all except https://github.com/libvips/libvips/pull/1346#discussion_r298503723

**Comment by jcupitt** [comment]
Great! I should have some time this weekend, I will reread all this.

File: libvips/foreign/gifload.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcupitt) [reply libvips/foreign/gifload.c]: No problem, I'll have a look at this again after merge.

**Comment by jcupitt** [comment]
OK, all merged! Thank you again!

I made some minor formatting changes and revised a couple of loops. I added some tests too. It all valgrinds cleanly as well.

**Comment by deftomat** [comment]
AWESOME! Thanks.

**Pull request merged**